src/upgrader.ml

```diff
@@ -0,0 +1,243 @@
+    | Quoted_string (loc, _) as x ->
+      Dune_lang.Decoder.parse String_with_vars.decode
+        (Univ_map.singleton (Syntax.key Stanza.syntax) (0, 0))
+        x
+      |> String_with_vars.upgrade_to_dune ~allow_first_dep_var:true
+      |> String_with_vars.encode
+      |> Dune_lang.add_loc ~loc
+    | List (loc, l) ->
+      let l =
+        match l with
+        | [Atom _; List (_, [Atom (_, A ":include"); Atom _])] ->
+          List.map l ~f:upgrade
+        | [Atom (_, A field_name) as field; List (_, args)]
+          when simplify_field field_name ->
+          upgrade field :: List.map args ~f:upgrade
+        | _ ->
+          List.map l ~f:upgrade
+      in
+      let l =
+        if List.exists l ~f:uses_first_dep_var then
+          List.map l ~f:(function
+            | List (loc, (Atom (_, A "deps") as field :: first :: rest)) ->
+              List (loc,
+                    field ::
+                    (let loc = Dune_lang.Ast.loc first in
+                     List (loc,
+                           [Atom (loc, Dune_lang.Atom.of_string ":<");
+                            first])) ::
+                    rest)
+            | x -> x)
+        else
+          l
+      in
+      List (loc, l)
+  in
+  upgrade stanza
+
+let upgrade_file todo file sexps ~look_for_jbuild_ignore =
+  let dir = Path.parent_exn file in
+  let new_file =
+    let base = Path.basename file in
+    let new_base =
+      match String.drop_prefix base ~prefix:"jbuild" with
+      | None -> base
+      | Some suffix -> "dune" ^ suffix
+    in
+    Path.relative dir new_base
+  in
+  let sexps =
+    List.filter sexps ~f:(function
+      | Dune_lang.Ast.List (_, [Atom (_, A "jbuild_version"); _]) -> false
+      | _ -> true)
+  in
+  let sexps = List.map sexps ~f:upgrade_stanza in
+  let sexps, extra_files_to_delete =
+    (* Port the jbuild-ignore file if necessary *)
+    let jbuild_ignore = Path.relative dir "jbuild-ignore" in
+    if not (look_for_jbuild_ignore && Path.exists jbuild_ignore) then
+      (sexps, [])
+    else begin
+      let data_only_dirs = File_tree.load_jbuild_ignore jbuild_ignore in
+      let stanza =
+        Dune_lang.add_loc ~loc:Loc.none
+          (List (Dune_lang.atom "data_only_dirs"
+                 :: List.map (String.Set.to_list data_only_dirs)
+                      ~f:Dune_lang.atom_or_quoted_string))
+      in
+      let sexps = stanza :: sexps in
+      (sexps, [jbuild_ignore])
+    end
+  in
+  let contents = Format.asprintf "%a@?" Dune_fmt.pp_top_sexps sexps in
+  todo.to_rename_and_edit <-
+    { original_file = file
+    ; new_file
+    ; extra_files_to_delete
+    ; contents
+    } :: todo.to_rename_and_edit
+
+let upgrade_dir todo dir =
+  let project = File_tree.Dir.project dir in
+  (match Dune_project.ensure_project_file_exists project with
+   | Already_exist -> ()
+   | Created ->
+     let project_file =
+       Path.relative (Path.of_local (Dune_project.root project))
+         Dune_project.filename
```

Comment:
It might simpler to have a `Dune_project.path : t -> Path.t` function that returns the path of the dune project.

Reply:
Sounds good, I name it just `Dune_project.file`.